proposals/002-Auttomaticly-reload-on-Kubernetes-config-change.md

```diff
@@ -0,0 +1,75 @@
+      1. Compute a checksum of secret UID & Generation
+3. Annotate the KafkaService with `kroxylicious.io/dependants-checksum = <computedChecksum>`  
+
+
+[^1]: For simplicity CRC32 as that's already supported by the JDk. However, the algorithm used is immaterial.
+## Advantages
+Maintains the same logical division between reconcilers thus avoiding the need for the aggregating reconciler to watch all resources involved and thus the need for it to understand the details of the full configuration model and its kubernetes representation. 
+
+## Limitations
+1. The blast radius of any config change is still the whole proxy process. I.e. a malformed config file still prevents the whole proxy from starting.
+This is largely mitigated by the fact the config file is in the kubernetes case is programmatically generated, and we can validate that before applying it.
+2. Any change to the config model requires the entire proxy process to restart. I.e. renewing the certificate of one virtual cluster requires the all connections to the hosting proxy instance to be disconnected and reconnect.
+3. The operator requires permissions to actually get all the resources (in order to obtain the `metadata.generation` and `metadata.uid`). While it can just query for the specific metadata the Kubernetes RBAC model works at verb/resource level not to fields within resources. 
+
+# Rejected Alternatives
+1. Have the Kafka Proxy reconciler (KPR) watch all resources involved for changes and trigger on any update. This has a variety of issues
+   1. We conflate concerns between reconcilers as we end up with two (or more) reconcilers watching the same resource.
+   2. The KPR can, and likely will, be triggered on invalid or intermediate states thus making for vastly more complicated error handling and recovery
+   3. It becomes much harder for operators to work out what triggered a particular change to the running processes.
+2. The proxy process monitors the config file(s) for changes 
+   1. What set of files should the proxy monitor? The set can grow and shrink during parsing as filers are added and removed
+   2. How does the proxy know when a file is fully updated? User saves changes to the live config file before looking up the correct config syntax.
+   3. We add complexity to the proxy, extra threads for file watches.
+   4. There is no mechanism for the proxy to report errors back to the operator. e.g. conflicting port allocations. 
+3. Follow the model pioneered by HTTPD but copied by many other services which support vhosting of a `conf.d` directory and have one file per virtual cluster.
+   1. Increases the complexity of managing bare metal clusters 
+   2. Makes crafting valid config harder (identifying for port collisions across a directory of file)
+   3. Increases complexity within the proxy
+   4. No mechanism to report problems encountered at the proxy level back to the operator.
+   5. Lack of developer bandwidth to context switch to the proxy
```

Comment:
I'm not sure what this bullet means.

Reply:
I was trying to say we are all focused on the operator and we don't have the capacity to go re-work proxy config loading as otherwise this model is really a big part of the answer. 

Actually maybe its not, this addresses the orthogonal concern of blast radius, tweaking on VKC causing a stampeding heard as all connections restart. It doesn't really address the co-ordination problem